src/backend/mpi/swdsm.cpp

```diff
@@ -436,315 +433,264 @@ void handler(int sig, siginfo_t *si, void *unused){
+					storepageDIFF(idx+j,pagesize*j+(cacheControl[idx].tag));
 				}
-
-				cacheControl[startidx].state = INVALID;
-				cacheControl[startidx].tag = lineAddr;
-
-				cacheControl[startidx].dirty=CLEAN;
-				vm::map_memory(lineptr, blocksize, pagesize*startidx, PROT_NONE);
-				mprotect(tmpptr2,blocksize,PROT_NONE);
+				argo_write_buffer->erase(idx);
 			}
+			/* Ensure the writeback has finished */
+			for(int i = 0; i < numtasks; i++){
+				if(barwindowsused[i] == 1){
+					MPI_Win_unlock(i, globalDataWindow[i]);
+					barwindowsused[i] = 0;
+				}
+			} // TODO: move this out to avoid unlocking for every page?
+
+			/* Clean up cache and protect memory */
+			cacheControl[idx].state = INVALID;
+			cacheControl[idx].tag = temp_addr;
+			cacheControl[idx].dirty = CLEAN;
+			vm::map_memory(temp_ptr, block_size, pagesize*idx, PROT_NONE);
+			mprotect(old_ptr,block_size,PROT_NONE);
 		}
 	}
 
+	/* Initialize classification_index_array */
+	for(std::size_t i = 0; i < fetch_size; i+=CACHELINE){
+		const std::size_t temp_addr = aligned_access_offset + i*block_size;
+		classification_index_array[i] = get_classification_index(temp_addr);
+	}
 
-
+	/* Increase stat counter as load will be performed */
 	stats.loads++;
-	unsigned long classidx = get_classification_index(lineAddr);
-	unsigned long tempsharer = 0;
-	unsigned long tempwriter = 0;
 
+	/* Get globalSharers info from local node and add self to it */
 	MPI_Win_lock(MPI_LOCK_SHARED, workrank, 0, sharerWindow);
-	unsigned long prevsharer = (globalSharers[classidx])&id;
-	MPI_Win_unlock(workrank, sharerWindow);
-	int n;
-	homenode = getHomenode(lineAddr);
-
-	if(prevsharer==0 ){ //if there is strictly less than two 'stable' sharers
-		MPI_Win_lock(MPI_LOCK_SHARED, homenode, 0, sharerWindow);
-		MPI_Get_accumulate(&id, 1, MPI_LONG, &tempsharer, 1, MPI_LONG,
-			homenode, classidx, 1, MPI_LONG, MPI_BOR, sharerWindow);
-		MPI_Get(&tempwriter, 1,MPI_LONG,homenode,classidx+1,1,MPI_LONG,sharerWindow);
-		MPI_Win_unlock(homenode, sharerWindow);
-	}
-
-	MPI_Win_lock(MPI_LOCK_EXCLUSIVE, workrank, 0, sharerWindow);
-	globalSharers[classidx] |= tempsharer;
-	globalSharers[classidx+1] |= tempwriter;
-	MPI_Win_unlock(workrank, sharerWindow);
-
-	unsigned long offset = getOffset(lineAddr);
-	if(isPowerOf2((tempsharer)&invid) && tempsharer != id && prevsharer == 0){ //Other private. but may not have loaded page yet.
-		unsigned long ownid = tempsharer&invid; // remove own bit
-		unsigned long owner = invalid_node; // initialize to failsafe value
-		for(n=0; n<numtasks; n++) {
-			if(1ul<<n==ownid) {
-				owner = n; //just get rank...
-				break;
+	for(std::size_t i = 0; i < fetch_size; i+=CACHELINE){
+		if(pages_to_load[i]){
+			/* Check local pyxis directory if we are sharer of the page */
+			local_sharers[i] = (globalSharers[classification_index_array[i]])&node_id_bit;
+			if(local_sharers[i] == 0){
+				sharer_bit_mask[i*2] = node_id_bit;
+				new_sharer = true; //At least one new sharer detected
```

Comment:
do I understand this correctly that even when we always load 8 pages, the coherence data is stored separately for each page? I'm not asking for this to change, just making sure I understand it correctly.

Reply:
We load _at most_ 8 pages, as there may be conditions that prevent sequential reads of `ARGO_LOAD_SIZE` pages.

The coherence data is stored separately for each page in the pyxis directory (`globalSharers`), with one `unsigned long` storing sharers and one storing writers.